include/kernels/coarse_op_kernel.cuh

```diff
@@ -120,74 +128,176 @@ namespace quda {
+    case 1: return arg.comm_dim[1] && coord[1] + nFace >= arg.x_size[1];
+    case 2: return arg.comm_dim[2] && coord[2] + nFace >= arg.x_size[2];
+    case 3: return arg.comm_dim[3] && coord[3] + nFace >= arg.x_size[3];
+    }
+    return false;
+  }
+
+  /**
+     @brief Helper for computing checkerboard index from from a shift
+     of length nFace in the direction dim
+     @param[in] x Grid coordinates
+     @param[in] X Grid dimensions
+     @param[in] dim Dimension of the shift
+     @param[in] nFace Depth of the halo
+  */
+  template <typename I, typename Coord>
+  __device__ __host__ inline auto linkIndexHop(const Coord &x, const I X[4], const int dim, int nFace)
+  {
+    int y[4];
+#pragma unroll
+    for ( int i = 0; i < 4; i++ ) y[i] = x[i];
+    switch (dim) {
+    case 0: y[0] = (y[0] + nFace + X[0]) % X[0]; break;
+    case 1: y[1] = (y[1] + nFace + X[1]) % X[1]; break;
+    case 2: y[2] = (y[2] + nFace + X[2]) % X[2]; break;
+    case 3: y[3] = (y[3] + nFace + X[3]) % X[3]; break;
+    }
+    return (((y[3] * X[2] + y[2]) * X[1] + y[1]) * X[0] + y[0]) >> 1;
+  }
 
+  /**
+     @brief Helper for computing if we are on the edge of an aggregate
+     @param[in] coord Fine grid coordinates
+     @param[in] coord_coarse Coarse grid coordinates
+     @param[in] dim Dimension
+     @param[in] arg Kernel argument
+  */
+  template <typename Arg> constexpr bool isCoarseDiagonal(const int coord[], const int coord_coarse[], int dim, const Arg &arg)
+  {
+    switch (dim) {
+    case 0: return ((coord[0] + 1) % arg.x_size[0]) / arg.geo_bs[0] == coord_coarse[0];
+    case 1: return ((coord[1] + 1) % arg.x_size[1]) / arg.geo_bs[1] == coord_coarse[1];
+    case 2: return ((coord[2] + 1) % arg.x_size[2]) / arg.geo_bs[2] == coord_coarse[2];
+    case 3: return ((coord[3] + 1) % arg.x_size[3]) / arg.geo_bs[3] == coord_coarse[3];
+    }
+    return false;
+  }
+
+  /**
+     @brief Calculates the matrix UV^{s,c'}_mu(x) = \sum_c U^{c}_mu(x) * V^{s,c}_mu(x+mu)
+     Where: mu = dim, s = fine spin, c' = coarse color, c = fine color
+     or, if dir == QUDA_IN_PLACE, UV^{s,c'}(x) = \sum_c C^{c}_mu(x) * V^{s,c}_mu(x+mu)
+     @param[in] arg Kernel argumnt
+     @param[in] Wacc Input vector accessor
+     @param[in] parity Parity index
+     @param[in] x_cb Checkerboard index
+     @param[in] i0 Color color row index (coarse)
+     @param[in] j0 Color column index (fine)
+  */
+  template <typename Wtype, typename Arg>
+  __device__ __host__ inline auto computeUV(const Arg &arg, const Wtype &Wacc, int parity, int x_cb, int i0, int j0)
+  {
     constexpr int uvSpin = Arg::fineSpin * (Arg::from_coarse ? 2 : 1);
     constexpr int nFace = 1; // to do: nFace == 3 version for long links
 
-    using complex = complex<typename Arg::Float>;
+    using real = typename Arg::Float;
+    using complex = complex<real>;
     using TileType = typename Arg::uvTileType;
     auto &tile = arg.uvTile;
+    auto dim = arg.dim;
```

Comment:
Use `arg.dim` elsewhere instead of `dim` for consistency with the replacement of `dir` -> `arg.dir`

Reply:
c94109308c46ca051eea11a865a2f445d990e257